[1.16.x] Made Grindstone and Anvil Recipe Type and Serializer.

ferriarnus opened the pull request:
Added a Grindstone recipe type and (default) serializer, making it possible to use json recipes and custom serializers for the grindstone. The grindstone first checks for a recipe, after which it does its default behaviour.  The grindstone now also accepts any item as input, not just damageable or book-items. The implemented serializer is a copy of the smithing table, but I will add nbt support. An example mod which adds a recipe is also included. If you have some comments, suggestions or concerns please tell them.

ferriarnus committed:
Initial commit

Added a Grindstone recipe type and (default) serializer, making it possible to use json recipes and custom serializers for the grindstone

Thelnfamous1 commented:
I like the idea. If you can figure out the NBT stuff for grinding, perhaps add a recipe type for the anvil as well if the logic is similar enough. It’d be extremely nice to use both of these containers for some data-driven NBT crafting.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/inventory/container/GrindstoneContainer.java.patch:
```diff
@@ -1,5 +1,46 @@
+    public GrindstoneContainer(int p_i50081_1_, PlayerInventory p_i50081_2_, final IWorldPosCallable p_i50081_3_) {
+       super(ContainerType.field_221521_o, p_i50081_1_);
+       this.field_217015_e = p_i50081_3_;
++      this.level = p_i50081_2_.field_70458_d.field_70170_p;
```
You have an `IWorldPosCallable`, this addition is unnecessary.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/inventory/container/GrindstoneContainer.java.patch:
```diff
@@ -1,5 +1,46 @@
++      this.level = p_i50081_2_.field_70458_d.field_70170_p;
+       this.func_75146_a(new Slot(this.field_217014_d, 0, 49, 19) {
+          public boolean func_75214_a(ItemStack p_75214_1_) {
+-            return p_75214_1_.func_77984_f() || p_75214_1_.func_77973_b() == Items.field_151134_bR || p_75214_1_.func_77948_v();
++            return true;
```
Breaks vanilla behavior. This is my opinion, but you should check whether the stack is valid rather than just saying all stacks are valid. This could be justified in either direction however.

ferriarnus replied:
How would I go about checking? I can't call the recipe if I don't have both items to check. Are the recipe items stored somewhere (like a list)?

ChampionAsh5357 replied:
No, but you can construct and invalidate a map for that. But again, this starts being overengineered and questions whether Forge should have it since this checking mechanic could be justified for or against depending on how much is implemented.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/inventory/container/GrindstoneContainer.java.patch:
```diff
@@ -1,5 +1,46 @@
+       });
+       this.func_75146_a(new Slot(this.field_217014_d, 1, 49, 40) {
+          public boolean func_75214_a(ItemStack p_75214_1_) {
+-            return p_75214_1_.func_77984_f() || p_75214_1_.func_77973_b() == Items.field_151134_bR || p_75214_1_.func_77948_v();
++            return true;
```
Breaks vanilla behavior. This is my opinion, but you should check whether the stack is valid rather than just saying all stacks are valid. This could be justified in either direction however.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/inventory/container/GrindstoneContainer.java.patch:
```diff
@@ -1,5 +1,46 @@
+       ItemStack itemstack1 = this.field_217014_d.func_70301_a(1);
+       boolean flag = !itemstack.func_190926_b() || !itemstack1.func_190926_b();
+       boolean flag1 = !itemstack.func_190926_b() && !itemstack1.func_190926_b();
++      java.util.Optional<net.minecraftforge.common.crafting.GrindStoneRecipe> optional = this.level.func_199532_z().func_215371_a(net.minecraft.item.crafting.IRecipeType.GRINDING, this.field_217014_d, this.level);
++      if (optional.isPresent()) {
++        net.minecraftforge.common.crafting.GrindStoneRecipe grindStoneRecipe = optional.get();
++        ItemStack result = grindStoneRecipe.func_77572_b(this.field_217013_c);
++        this.field_217013_c.func_70299_a(0, result);
++        this.func_75142_b();
++        return;
++      }
```
This can be a one line patch with an external method reference. Also, it's located in the wrong place. The flag after your patch checks to see if both are empty so this call executes unnecessarily. You should either move it where it needs to go or check flag1 and execute the method check. That way no unnecessary evaluations are done.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/item/crafting/IRecipeSerializer.java.patch:
```diff
@@ -9,7 +9,11 @@
+    CookingRecipeSerializer<CampfireCookingRecipe> field_222174_r = func_222156_a("campfire_cooking", new CookingRecipeSerializer<>(CampfireCookingRecipe::new, 100));
+    IRecipeSerializer<StonecuttingRecipe> field_222175_s = func_222156_a("stonecutting", new SingleItemRecipe.Serializer<>(StonecuttingRecipe::new));
+    IRecipeSerializer<SmithingRecipe> field_234826_u_ = func_222156_a("smithing", new SmithingRecipe.Serializer());
++   IRecipeSerializer<net.minecraftforge.common.crafting.GrindStoneRecipe> GRINDSTONE = func_222156_a("grinding", new net.minecraftforge.common.crafting.GrindStoneRecipe.Serializer());
```
This is not how you register within forge. See `ForgeMod` and how attributes are registered. You'll find its the same as the documentation.

ferriarnus replied:
Thank you, I was also wondering about that. This makes it a "vanilla" method, which it actually isn't. Will take a look at `ForgeMod`

ChampionAsh5357 commented on patches/minecraft/net/minecraft/item/crafting/IRecipeType.java.patch:
```diff
@@ -0,0 +1,10 @@
+    IRecipeType<CampfireCookingRecipe> field_222153_e = func_222147_a("campfire_cooking");
+    IRecipeType<StonecuttingRecipe> field_222154_f = func_222147_a("stonecutting");
+    IRecipeType<SmithingRecipe> field_234827_g_ = func_222147_a("smithing");
++   IRecipeType<net.minecraftforge.common.crafting.GrindStoneRecipe> GRINDING = func_222147_a("grinding");
```
Unnecessary patch, move to `ForgeMod`.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+package net.minecraftforge.common.crafting;
```
Missing license

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+public class GrindStoneRecipe implements IRecipe<IInventory>
+{
+
+	private ResourceLocation id;
```
Forge uses four spaces, not tabs.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+
+public class GrindStoneRecipe implements IRecipe<IInventory>
```
No documentation and classname does not match vanilla naming scheme.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+	private ItemStack result;
+	
+
+	public GrindStoneRecipe(ResourceLocation id,Ingredient ingredient, Ingredient ingredient1, ItemStack result) 
```
Inconsistent parameter formatting

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+	}
+	
+	@Override
+	public boolean matches(IInventory p_77569_1_, World p_77569_2_) 
```
Param names in your own class, rename to something understandable.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+		return IRecipeType.GRINDING;
+	}
+	
+	public static class Serializer extends net.minecraftforge.registries.ForgeRegistryEntry<IRecipeSerializer<?>> implements IRecipeSerializer<GrindStoneRecipe> 
```
Does this really need to be fully qualified if it's your own class?

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+		public GrindStoneRecipe fromNetwork(ResourceLocation p_199426_1_, PacketBuffer p_199426_2_) 
+		{
+			Ingredient ingredient = Ingredient.fromNetwork(p_199426_2_);
+	        Ingredient ingredient1 = Ingredient.fromNetwork(p_199426_2_);
```
Mixture of spaces and tabs everywhere.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+		}
+		
+	}
+
+	
+	
+
```
Weird blank space.

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/item/crafting/GrindStoneRecipeTest.java:
```diff
@@ -0,0 +1,35 @@
+import net.minecraftforge.fml.common.Mod;
+
+@Mod(GrindStoneRecipeTest.MOD_ID)
+@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD, modid = GrindStoneRecipeTest.MOD_ID)
```
But there's no events?

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+	{
+		this.id = id;
+		this.ingredient = ingredient;
+		this.ingredient1 = ingredient1;
```
What is ingredient1?

ferriarnus replied:
The grindstone has 2 input slots, so their are 2 ingredients. I will probably change it to a list, similar to Shapeless, unless someone thinks its better to keep them separate? Than I will find some better names.

ChampionAsh5357 replied:
Yes, but those are not the name of the slots you gave it, you called it `base` and `addition`, which does not match the naming scheme here. Plus, they're just two generic ingredients as the grindstone could be considered a shapeless recipe with only two slots. It doesn't matter which goes in which. So, actually you'll probably need to reimplement the checking logic as well.

ferriarnus replied:
Yes I agree, like I said, I copied the smithing table for it, which uses those tags. I will change it to match what it is for the grindstone.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindStoneRecipe.java:
```diff
@@ -0,0 +1,110 @@
+	public GrindStoneRecipe(ResourceLocation id,Ingredient ingredient, Ingredient ingredient1, ItemStack result) 
+	{
+		this.id = id;
+		this.ingredient = ingredient;
```
What is ingredient?

ChampionAsh5357 commented on src/test/java/net/minecraftforge/debug/item/crafting/GrindStoneRecipeTest.java:
```diff
@@ -0,0 +1,35 @@
+@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD, modid = GrindStoneRecipeTest.MOD_ID)
+public class GrindStoneRecipeTest
+{
+    public static final String MOD_ID = "grindstonerecipe_test";
```
Nitpick, but why one word except for `test`?

ChampionAsh5357 commented on src/test/resources/data/grindstonerecipes_test/recipes/test.json:
```diff
@@ -0,0 +1,12 @@
+  "result": {
+    "item": "minecraft:netherite_axe"
+  }
+}
```
Inconsistent line ending.

ChampionAsh5357 commented on src/test/resources/data/grindstonerecipes_test/recipes/test.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "type": "minecraft:grinding",
```
This is not a minecraft added recipe. Should be under the forge namespace.

ChampionAsh5357 reviewed (commented):
This is missing some tools that would be expected of a new recipe type. Specifically, the data providers for creating this recipe and a test of it working. That would also need to have the forge nbt support as is provided within the grindstone recipe deserializer. There are also naming schemes which indicate no verbosity as to what the field or parameter is for.

I'm not sure how valid of a PR this is. One of the major points that comes to mind is the recipe book. A good portion of vanilla recipes use the recipe book. And for those that don't, the recipes are available to the user in a different format. Only case this does not hold true is for the brewing recipes but they use a different recipe system. However, this PR forces the grindstone to conform to the same behavior and logic that vanilla would follow. So would there be a recipe book implementation? Would it have some other way to show recipes? Or would it just completely lack this feature since vanilla did not intend this to be data driven (yet?). 

Another opinion is on the anvil which is essentially the same as a grindstone. Why not do something similar with an event system? Should the anvil also conform to this data driven behavior? If so, will the above concerns also be addressed.

This system more or less just seems like a patchwork feature rather than a fully implemented patchwork feature. This is my opinion on the matter anyways. It might be completely valid and my concerns are just overly thought. But it does bring up the point of how much should Forge patch to implement data driven features and which other recipe systems should be patched to implement as such.

ferriarnus commented:
It's still a draft cause i knew there would be quiet a few issue I couldn't solve without feedback. Adding the anvil is a good idea and I was planning to add it to this PR if the way I did it here worked. So the main problem (outside of misformatted code and poor naming) is if the grindstone (and anvil) should get a "vanilla" based recipe type. What do you mean with "event"? Add an event when the grindstone or anvil is used? This works, but would make them completely different from "normal". This may be the right move, because they are different, but I think it would confuse people. It is also easier to use and add a lot of recipes if it uses the json files (when I a add a data provider). But this is not my decision to make. I will clean up the code and hopefully Forge can make a decision on if it's valid (or what would be needed to make it valid). The recipe book is a good point, but honestly, not sure if it's used enough by the people to warrant a problem. JEI is widely used and mostly unmissable in any modded minecraft world. The grindstone could be added to JEI (if the owner thinks it's good) to solve the recipe issue. But again, not sure what Forge's stance is on "essential" mods like this.

ferriarnus committed:
Used ForgeMod

Moved registration to the ForgeMod class. Fixed tabs. Still working on nbt. Still working on cleaning up the grindstonecontainer.

ChampionAsh5357 commented:
> What do you mean with "event"?

Look at `AnvilUpdateEvent`. It  is Forge's current addition for a recipe system within anvils.

> JEI is widely used and mostly unmissable in any modded minecraft world. The grindstone could be added to JEI (if the owner thinks it's good) to solve the recipe issue.

This doesn't really solve the recipe issue. All it does is says to view recipes, a feature which vanilla adds natively, ForgeMod must rely on JEI to which JEI relies on ForgeMod. It's a circular dependency construct even though Forge suggests using JEI.

ferriarnus commented:
The issue with an event as well is that the slots don't accept items. You said it before (and that its even a breaking change, though I'm not certain it's that severe, as vanilla still works), the slots in the gui will need to be patched to take other items. And deciding which is not easy, if you want only items used in recipes.

ferriarnus committed:
Moved get recipe to other function

Moved the part which gets a recipe to the slotsChanged functions, making it be called after all vanilla code

ChampionAsh5357 commented:
> And deciding which is not easy, if you want only items used in recipes.

It doesn't break vanilla results, it just breaks expected behavior where you expect the grindstone to only allow items that it can actually grind. But again, this is really nuance so the justification for just allowing any item or only specific items is either or.

ferriarnus committed:
fix patch mistake?

ferriarnus committed:
added packet filter

Added a filter to make it possible to connect to a vanilla client. Also changed the names of the ingredients to match the smiting table. I'm thinking of keeping them "shaped". A modder could add a shapeless serialzer (or just add it twice).

ferriarnus committed:
Made provider

Made a data provider for grinding recipe. Moved the crafting call to a new helper class. Added check for "illegal" item, to compensate allowing them in the slots.

ferriarnus committed:
update test mod

ferriarnus commented:
I think everything got fixed now (or at least all I could see). Is everything acceptable? Can/should I start on the anvil, or is it better to wait a bit for others to pick into the discussion of adding recipe types vs events?

ferriarnus committed:
Update GrindingRecipe.java

ferriarnus commented:
So it's still a breaking change rigth? Should I port it to 1.17.1? Are there still issues?

ChampionAsh5357 commented on patches/minecraft/net/minecraft/network/play/server/SUpdateRecipesPacket.java.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+    }
+ 
+-   @OnlyIn(Dist.CLIENT)
++   //@OnlyIn(Dist.CLIENT)
```
I'm pretty sure you can't do this.

KnightMiner replied:
Forge has a [side only annotation stripper](https://github.com/MinecraftForge/MinecraftForge/blob/1.17.x/src/main/resources/forge.sas), I don't think commenting out only in is the way they do it

ferriarnus replied:
Oh, didn't know it existed. I will change it.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/gui/GuiHelper.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Adds Grinding recipes to the grinder. Takes the container, the IWorldPosCallable, the input and output inventory.
+     */
+    public static void grindingHelper(GrindstoneContainer container, IWorldPosCallable access, IInventory input, IInventory output)
```
Why couldn't the method go within `ForgeHooksClient` or some existing gui helper (pretty sure there is one somewhere)?

ferriarnus replied:
I wasn't sure where it needed to go. Will move it

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/gui/GuiHelper.java:
```diff
@@ -0,0 +1,56 @@
+public class GuiHelper
+{
+
+    /**
+     * Adds Grinding recipes to the grinder. Takes the container, the IWorldPosCallable, the input and output inventory.
+     */
```
Javadoc is incorrect, missing params

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/gui/GuiHelper.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    public static void grindingHelper(GrindstoneContainer container, IWorldPosCallable access, IInventory input, IInventory output)
+    {
+        if (input.getItem(0).isEmpty() && input.getItem(1).isEmpty())
```
This would be an or, both need to be present for a grindstone recipe.

ferriarnus replied:
But there could be recipes with only one item (like gemcutting or something)

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/gui/GuiHelper.java:
```diff
@@ -0,0 +1,56 @@
+        }
+        access.execute((level, pos) -> {
+            Optional<GrindingRecipe> optional = level.getRecipeManager().getRecipeFor(ForgeMod.GRINDING, input, level);
+            if (optional.isPresent()) {
+                GrindingRecipe grindingRecipe = optional.get();
+                ItemStack result = grindingRecipe.assemble(input);
+                output.setItem(0, result);
+                container.broadcastChanges();
+            }
```
I would use `#ifPresent` over this.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/ForgeMod.java:
```diff
@@ -112,7 +114,9 @@
     public static final RegistryObject<Fluid> FLOWING_MILK = RegistryObject.of(new ResourceLocation("flowing_milk"), ForgeRegistries.FLUIDS);
-
+    
+    public static final IRecipeType<GrindingRecipe> GRINDING = new IRecipeType<GrindingRecipe>() {};
```
Since Forge is against static initialization, I would probably leave this null to be set during that particular phase, but this would need to be confirmed by someone on the triage or forge team since most of them usually just static init recipe types.

alcatrazEscapee replied:
~~Wait a minute this shouldn't be `new IRecipeType<>` in any case, it should be registering it (`IRecipeType#register`). (Notably, the registry in question is barely used, but still, there's a register method *right there*.~~ Nevermind, I found the registration~~

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/crafting/GrindingRecipe.java:
```diff
@@ -0,0 +1,120 @@
+    private Ingredient addition;
+    private ItemStack result;
+    
+    public GrindingRecipe(ResourceLocation id,Ingredient ingredient, Ingredient ingredient1, ItemStack result) 
```
Once again with the mismatched param and field names. Make a decision to rename one or the other.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/network/VanillaConnectionNetworkFilter.java:
```diff
@@ -100,4 +104,21 @@ private static SCommandListPacket filterCommandList(SCommandListPacket packet)
+     * A vanilla client would fail to deserialize the packet and disconnect with an error message if these were sent.
+     */
+    @Nonnull
+    private static SUpdateRecipesPacket filterRecipes(SUpdateRecipesPacket packet)
```
You'll probably need to filter out advancements as well.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/data/GrindingRecipeBuilder.java:
```diff
@@ -0,0 +1,141 @@
+    private final Advancement.Builder advancement = Advancement.Builder.advancement();
+    private final IRecipeSerializer<?> type;
+    
+    public GrindingRecipeBuilder(IRecipeSerializer<?> type, Ingredient base, Ingredient addition, Item result)
```
Shouldn't be `Item` as data type, should serialize to `ItemStack` since that is supported.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/common/data/GrindingRecipeBuilder.java:
```diff
@@ -0,0 +1,141 @@
+            tag.add("base", this.base.toJson());
+            tag.add("addition", this.addition.toJson());
+            JsonObject jsonobject = new JsonObject();
+            jsonobject.addProperty("item", Registry.ITEM.getKey(this.result).toString());
+            tag.add("result", jsonobject);
```
Again, serialize to `ItemStack`.

ChampionAsh5357 reviewed (commented):
This is not a breaking change afaik. You're not modifying anything pre-existing, just adding new material. The patches don't change anything of consequence either besides where to inherit behavior from. This should be ported to 1.17 though and kept on 1.16 as well. Otherwise, here are my current comments.

ferriarnus committed:
small fixxes

changed Item to ItemStack, renamed some parameters, added advancement filter, moved function to new class

ferriarnus committed:
Update ForgeHooksClient.java

ferriarnus committed:
added anvil type

Added the anvil recipe type using the already existing event.

ferriarnus committed:
Update RepairContainer.java.patch

ferriarnus committed:
Merge branch '1.16.x' into GrindStoneRecipe

ferriarnus committed:
Update ForgeHooksClient.java

ferriarnus commented:
I've started adding the anvil (blacksmithing) recipe type. Since there was an event I simply used it to add it without a complicated patch. I still needed to change the experience cost of crafting, making it so you can actually craft items with no cost. I will start  on the data generators tomorrow. Any questions/remarks?

ferriarnus committed:
finished datagen

ferriarnus commented:
Everything should be in place right now, I think it's ready for review again.

ferriarnus committed:
changed grindstone helper

ferriarnus committed:
Delete IRecipeSerializer.java.patch

ferriarnus committed:
small fixes

ferriarnus committed:
Update IRecipeSerializer.java.patch

ferriarnus committed:
Update IRecipeSerializer.java.patch

Well I can't figure out how to remove this patch it seems

ferriarnus committed:
Patches to containers

Some patches to the grindstone and repair container needed to only take 1 item instead of all when crafting. Also update the slots after taking out an item

ferriarnus committed:
yet another container patch

ChampionAsh5357 commented:
Closing this since it has been determined as out of scope for Forge. See Curle's explanation on the [brewing datapack PR](https://github.com/MinecraftForge/MinecraftForge/pull/7949#issuecomment-944211848) for more details.
